src/functions/env.ts

```diff
@@ -0,0 +1,238 @@
+ *
+ *   PUBLIC_KEY="-----BEGIN PUBLIC KEY-----
+ *   ABC
+ *   EFG
+ *   -----BEGIN PUBLIC KEY-----"
+ *
+ * See test for more examples.
+ *
+ * @return {ParseResult} Result containing parsed key, value pairs and errored lines.
+ */
+export function parse(data: string): ParseResult {
+  const envs: Record<string, string> = {};
+  const errors: string[] = [];
+
+  data = data.replace(/\r\n?/, "\n"); // For Windows support.
+  let match;
+  while ((match = LINE_RE.exec(data))) {
+    let [, k, v] = match;
+    v = (v || "").trim();
+
+    let quotesMatch;
+    if ((quotesMatch = /^(["'])(.*)\1$/ms.exec(v)) != null) {
+      // Remove surrounding single/double quotes.
+      v = quotesMatch[2];
+      if (quotesMatch[1] === '"') {
+        // Unescape newlines and tabs.
+        v = v.replace("\\n", "\n").replace("\\r", "\r").replace("\\t", "\t").replace("\\v", "\v");
+        // Unescape other escapable characters.
+        v = v.replace(/\\([\\'"])/g, "$1");
+      }
+    }
+    envs[k] = v;
+  }
+
+  const nonmatches = data.replace(LINE_RE, "");
+  for (let line of nonmatches.split(/[\r\n]+/)) {
+    line = line.trim();
+    if (line.startsWith("#")) {
+      // Ignore comments
+      continue;
+    }
+    if (line.length) errors.push(line);
+  }
+
+  return { envs, errors };
+}
+
+/**
+ * Validates string for use as an env var key.
+ *
+ * We restrict key names to ones that conform to POSIX standards.
+ * This is more restrictive than what is allowed in Cloud Functions or Cloud Run.
+ */
+export function validateKey(key: string): void {
+  if (RESERVED_KEYS.includes(key)) {
+    throw new Error("Key reserved for internal use.");
+  }
+  if (!/^[A-Z_][A-Z0-9_]*$/.test(key)) {
+    throw new Error(
+      "Key must start with an uppercase ASCII letter or underscore" +
+        ", and then consist of uppercase ASCII letters, digits, and underscores."
+    );
+  }
+  if (key.startsWith("X_GOOGLE_") || key.startsWith("FIREBASE_")) {
+    throw new Error("Key starts with a reserved prefix (X_GOOGLE_ or FIREBASE_)");
+  }
+}
+
+// Parse dotenv file, but throw errors if:
+//   1. Input has any invalid lines.
+//   2. Any env key is reserved.
+function parseStrict(data: string): Record<string, string> {
+  const { envs, errors } = parse(data);
+
+  if (errors.length) {
+    logger.debug(`Invalid dotenv file, error on lines:\n${errors.join("\n")}`);
+    throw new Error(`Invalid dotenv file, error on lines: ${errors.join(",")}`);
+  }
+
+  const invalidKeys: Record<string, string> = {};
+  for (const key of Object.keys(envs)) {
+    try {
+      validateKey(key);
+    } catch (err) {
+      invalidKeys[key] = err.message || "unknown error";
+    }
+  }
+  if (Object.keys(invalidKeys).length > 0) {
+    logger.debug(
+      `Found invalid keys: ${Object.entries(invalidKeys)
+        .map(([k, v]) => `${k}: ${v}`)
+        .join("\n")}`
+    );
+    throw new Error(`Found invalid keys: ${Object.keys(invalidKeys).join(",")}`);
```

Comment:
Again, I would consider
1. Using Firebase Error
2. Using the key name in the error message
3. Using FirebaseError.children instead of a new message here (assuming it shows up in the console correctly)

Reply:
Updated to use `children` prop. Like the result.